<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Parallelized calculation of split training/test set predictions from L1/L2/Elastic
  Net penalized regression. — opt.splitval • pensim</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Parallelized calculation of split training/test set predictions from L1/L2/Elastic
  Net penalized regression. — opt.splitval"><meta property="og:description" content="uses a single training/test split to train a penalized
  regression model in the training samples, then use the model to
  calculate values of the linear risk score in the test samples.  This
  function is used by opt.nested.crossval, but can also be used on its own.
This function support z-score scaling of training data, and application
  of these scaling and shifting coefficients to the test data.  It also
  supports repeated tuning of the penalty parameters and selection of the
  model with greatest cross-validated likelihood."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pensim</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/pensim.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/waldronlab/pensim/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Parallelized calculation of split training/test set predictions from L1/L2/Elastic
  Net penalized regression.</h1>
    
    <div class="hidden name"><code>opt.splitval.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>uses a single training/test split to train a penalized
  regression model in the training samples, then use the model to
  calculate values of the linear risk score in the test samples.  This
  function is used by opt.nested.crossval, but can also be used on its own.</p>
<p>This function support z-score scaling of training data, and application
  of these scaling and shifting coefficients to the test data.  It also
  supports repeated tuning of the penalty parameters and selection of the
  model with greatest cross-validated likelihood.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">opt.splitval</span><span class="op">(</span>optFUN<span class="op">=</span><span class="st">"opt1D"</span>,testset<span class="op">=</span><span class="st">"equal"</span>,scaling<span class="op">=</span><span class="cn">TRUE</span>,<span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>optFUN</dt>
<dd><p>"opt1D" for Lasso or Ridge regression, "opt2D" for Elastic Net.  See
    the help pages for these functions for additional arguments.</p></dd>

  <dt>testset</dt>
<dd><p>For the opt.splitval function ONLY.
    "equal" for randomly assigned equal training and test sets, or an
    integer vector defining the positions of the test samples in the
    response, penalized, and unpenalized arguments which are passed to
    the optL1, optL2, or cvl functions of the penalized R package.</p></dd>

  <dt>scaling</dt>
<dd><p>If TRUE, each feature (column) of the training samples (in
    matrix/dataframe specified by the penalized argument) are scaled to
    z-scores, then these scaling and shifting factors are applied to the
    test data.
    If FALSE, no scaling is done.</p></dd>

  <dt>...</dt>
<dd><p>Additional arguments are required, to be passed to the
    optL1 or optL2 function of the penalized R package.  See those help
    pages, and it may be desirable to test these arguments directly on
    optL1 or optL2 before using this more CPU-consuming and complex
    function.</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>This function does split sample model training and testing for a
  single split of the data, using the optL1 or optL2 functions
  of the penalized R package, for each
  iteration of the cross-validation.  Scaling of the test samples is
  done independently, using scale factors determined from the training
  samples.  Repeated starts of model training can be parallelized as
  documented in the opt1D and opt2D functions.  This function is used
  for nested cross-validation by the opt.nested.crossval function.</p>
    </div>
    <div id="value">
    <h2>Value</h2>
    

<p>Returns a vector of cross-validated continuous risk score predictions.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Waldron L, Pintilie M, Tsao M-S, Shepherd FA, Huttenhower C*, Jurisica
  I*: Optimized application of penalized regression methods to diverse
  genomic data. Bioinformatics 2011, 27:3399-3406.  (*equal contribution)</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Levi Waldron et al.</p>
    </div>
    <div id="note">
    <h2>Note</h2>
    <p>Depends on the R packages: penalized, parallel, rlecuyer</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>opt1D, opt2D, opt.nested.crossval</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">beer.exprs</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">beer.survival</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## select just 250 genes to speed computation:</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">beer.exprs.sample</span> <span class="op">&lt;-</span> <span class="va">beer.exprs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">beer.exprs</span><span class="op">)</span>, <span class="fl">250</span><span class="op">)</span>, <span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gene.quant</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">beer.exprs.sample</span>, <span class="fl">1</span>, <span class="va">quantile</span>, probs <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dat.filt</span> <span class="op">&lt;-</span> <span class="va">beer.exprs.sample</span><span class="op">[</span><span class="va">gene.quant</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,<span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">gene.iqr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">dat.filt</span>, <span class="fl">1</span>, <span class="va">IQR</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dat.filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">dat.filt</span><span class="op">[</span><span class="va">gene.iqr</span> <span class="op">&gt;</span> <span class="fl">0.5</span>,<span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dat.filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">dat.filt</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival" class="external-link">survival</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">surv.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">beer.survival</span><span class="op">$</span><span class="va">os</span>, <span class="va">beer.survival</span><span class="op">$</span><span class="va">status</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Single split training/test evaluation.  Ideally nsim would be 50 and</span></span></span>
<span class="r-in"><span><span class="co">## fold=10, but this requires 100x more resources.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">preds50</span> <span class="op">&lt;-</span> <span class="fu">opt.splitval</span><span class="op">(</span></span></span>
<span class="r-in"><span>  optFUN <span class="op">=</span> <span class="st">"opt1D"</span>,</span></span>
<span class="r-in"><span>  scaling <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  testset <span class="op">=</span> <span class="st">"equal"</span>,</span></span>
<span class="r-in"><span>  setpen <span class="op">=</span> <span class="st">"L1"</span>,</span></span>
<span class="r-in"><span>  nsim <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>  nprocessors <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>  response <span class="op">=</span> <span class="va">surv.obj</span>,</span></span>
<span class="r-in"><span>  penalized <span class="op">=</span> <span class="va">dat.filt</span>,</span></span>
<span class="r-in"><span>  fold <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>  positive <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>  standardize <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>  trace <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">preds50.dichot</span> <span class="op">&lt;-</span> <span class="va">preds50</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">preds50</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">surv.obj.50</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>  <span class="va">surv.obj</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">preds50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">beer.survival</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">coxfit50.continuous</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/coxph.html" class="external-link">coxph</a></span><span class="op">(</span><span class="va">surv.obj.50</span> <span class="op">~</span> <span class="va">preds50</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">coxfit50.dichot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/coxph.html" class="external-link">coxph</a></span><span class="op">(</span><span class="va">surv.obj.50</span> <span class="op">~</span> <span class="va">preds50.dichot</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">coxfit50.continuous</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Call:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> coxph(formula = surv.obj.50 ~ preds50)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   n= 43, number of events= 13 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           coef exp(coef) se(coef)     z Pr(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> preds50  4.142    62.927    3.026 1.369    0.171</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         exp(coef) exp(-coef) lower .95 upper .95</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> preds50     62.93    0.01589    0.1671     23695</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Concordance= 0.618  (se = 0.084 )</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Likelihood ratio test= 1.84  on 1 df,   p=0.2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Wald test            = 1.87  on 1 df,   p=0.2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Score (logrank) test = 1.9  on 1 df,   p=0.2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">coxfit50.dichot</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Call:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> coxph(formula = surv.obj.50 ~ preds50.dichot)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   n= 43, number of events= 13 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     coef exp(coef) se(coef)     z Pr(&gt;|z|)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> preds50.dichotTRUE 0.623     1.865    0.560 1.113    0.266</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    exp(coef) exp(-coef) lower .95 upper .95</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> preds50.dichotTRUE     1.865     0.5363    0.6222     5.588</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Concordance= 0.561  (se = 0.071 )</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Likelihood ratio test= 1.24  on 1 df,   p=0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Wald test            = 1.24  on 1 df,   p=0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Score (logrank) test = 1.28  on 1 df,   p=0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Levi Waldron.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

