<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>pensim package user guide • pensim</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="pensim package user guide">
<meta property="og:description" content="pensim">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pensim</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/pensim.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/waldronlab/pensim/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>pensim package user guide</h1>
                        <h4 class="author">Levi Waldron</h4>
            
            <h4 class="date">March 27, 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/waldronlab/pensim/blob/master/vignettes/pensim.Rmd"><code>vignettes/pensim.Rmd</code></a></small>
      <div class="hidden name"><code>pensim.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This package acts as a wrapper to the <em>penalized</em> R package to add the following functionality to that package:</p>
<ul>
<li>repeated tuning on different data folds, with parallelization to take advantage of multiple processors</li>
<li>two-dimensional tuning of the Elastic Net</li>
<li>calculation of cross-validated risk scores through a nested cross-validation procedure, with parallelization to take advantage of multiple processors</li>
</ul>
<p>It also provides a function for simulation of collinear high-dimensional data with survival or binary response.</p>
<p>This package was developed in support of the study by <span class="citation">Waldron et al. (2011)</span>.<br>
This paper contains greater detail on proper application of the methods provided here. Please cite this paper when using the pensim package in your research, as well as the penalized package (<span class="citation">Goeman (2010)</span>).</p>
</div>
<div id="example-data" class="section level1">
<h1 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example data</h1>
<p><em>pensim</em> provides example data from a microarray experiment investigating survival of cancer patients with lung adenocarcinomas (<span class="citation">Beer et al. (2002)</span>). Load this data and do an initial pre-filter of genes with low IQR:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">pensim</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">beer.exprs</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">beer.survival</span>)
<span class="co">##select just 100 genes to speed computation, just for the sake of example:</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1</span>)
<span class="no">beer.exprs.sample</span> <span class="kw">&lt;-</span> <span class="no">beer.exprs</span>[<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">beer.exprs</span>), <span class="fl">100</span>),]
<span class="co">#</span>
<span class="no">gene.quant</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">beer.exprs.sample</span>, <span class="fl">1</span>, <span class="no">quantile</span>, <span class="kw">probs</span> <span class="kw">=</span> <span class="fl">0.75</span>)
<span class="no">dat.filt</span> <span class="kw">&lt;-</span> <span class="no">beer.exprs.sample</span>[<span class="no">gene.quant</span> <span class="kw">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(<span class="fl">100</span>),]
<span class="no">gene.iqr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">dat.filt</span>, <span class="fl">1</span>, <span class="no">IQR</span>)
<span class="no">dat.filt</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">dat.filt</span>[<span class="no">gene.iqr</span> <span class="kw">&gt;</span> <span class="fl">0.5</span>,])
<span class="no">dat.filt</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="no">dat.filt</span>)
<span class="no">dat.filt</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="no">dat.filt</span>)
<span class="co">#</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">survival</span>)
<span class="no">surv.obj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html">Surv</a></span>(<span class="no">beer.survival</span>$<span class="no">os</span>, <span class="no">beer.survival</span>$<span class="no">status</span>)</pre></body></html></div>
<p>Note that the expression data are in “wide” format, with one column per predictor (gene). It is recommended to put covariate data in a dataframe object, rather than a matrix.</p>
</div>
<div id="nested-cross-validation" class="section level1">
<h1 class="hasAnchor">
<a href="#nested-cross-validation" class="anchor"></a>Nested cross-validation</h1>
<p>Unbiased estimation of prediction accuracy involves two levels of cross-validation: an outer level for estimating prediction accuracy, and an inner level for model tuning. This process is simplified by the opt.nested.crossval function.</p>
<p>It is recommended first to establish the arguments for the penalized regression by testing on the <em>penalized</em> package functions <code>optL1</code> (for LASSO), <code>optL2</code> (for Ridge) or <code>cvl</code> (for Elastic Net). Here we use LASSO. Setting <code>maxlambda1=5</code> is not a generally recommended procedure, but is useful in this toy example to avoid converging on the null model.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">penalized</span>)</pre></body></html></div>
<pre><code>## Welcome to penalized. For extended examples, see vignette("penalized").</code></pre>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">testfit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/penalized/man/cvl.html">optL1</a></span>(
  <span class="kw">response</span> <span class="kw">=</span> <span class="no">surv.obj</span>,
  <span class="kw">penalized</span> <span class="kw">=</span> <span class="no">dat.filt</span>,
  <span class="kw">fold</span> <span class="kw">=</span> <span class="fl">5</span>,
  <span class="kw">maxlambda1</span> <span class="kw">=</span> <span class="fl">5</span>,
  <span class="kw">positive</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
  <span class="kw">standardize</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
  <span class="kw">trace</span> <span class="kw">=</span> <span class="fl">FALSE</span>
)</pre></body></html></div>
<p>Now pass these arguments to <code><a href="../reference/opt.nested.crossval.html">opt.nested.crossval()</a></code> for cross-validated calculation and assessment of risk scores, with the additional arguments:</p>
<ul>
<li>
<code>outerfold</code> and <code>nprocessors</code>: number of folds for the outer level of cross-validation, and the number of processors to use for the outer level of cross-validation (see <code><a href="../reference/opt.nested.crossval.html">?opt.nested.crossval</a></code>)</li>
<li>optFUN and scaling: which optimization function to use (<code>opt1D</code> for LASSO or Ridge, <code>opt2D</code> for Elastic Net) - see <code><a href="../reference/opt.splitval.html">?opt.splitval</a></code>.</li>
<li>setpen and nsim: setpen defines whether to do LASSO (“<strong>L1</strong>”) or Ridge (“<strong>L2</strong>”), <code>nsim</code> defines the number of times to repeat tuning (see <code><a href="../reference/opt1D.html">?opt1D</a></code>. <code>opt2D</code> has different required arguments.)</li>
</ul>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1</span>)
<span class="no">preds</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="../reference/opt.nested.crossval.html">opt.nested.crossval</a></span>(
    <span class="kw">outerfold</span> <span class="kw">=</span> <span class="fl">5</span>,
    <span class="kw">nprocessors</span> <span class="kw">=</span> <span class="fl">1</span>,
    <span class="co">#opt.nested.crossval arguments</span>
    <span class="kw">optFUN</span> <span class="kw">=</span> <span class="st">"opt1D"</span>,
    <span class="kw">scaling</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
    <span class="co">#opt.splitval arguments</span>
    <span class="kw">setpen</span> <span class="kw">=</span> <span class="st">"L1"</span>,
    <span class="kw">nsim</span> <span class="kw">=</span> <span class="fl">1</span>,
    <span class="co">#opt1D arguments</span>
    <span class="kw">response</span> <span class="kw">=</span> <span class="no">surv.obj</span>,
    <span class="co">#rest are penalized::optl1 arguments</span>
    <span class="kw">penalized</span> <span class="kw">=</span> <span class="no">dat.filt</span>,
    <span class="kw">fold</span> <span class="kw">=</span> <span class="fl">5</span>,
    <span class="kw">positive</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
    <span class="kw">standardize</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
    <span class="kw">trace</span> <span class="kw">=</span> <span class="fl">FALSE</span>
  )</pre></body></html></div>
<p>Ideally nsim would be 50, and outerfold and fold would be 10, but the values below speed computation 200x compared to these recommended values. Note that here we are using the <code>standardize=TRUE</code> argument of <code>optL1</code> rather than the <code>scaling=TRUE</code> argument of opt.splitval. These two approaches to scaling are roughly equivalent, but the scaling approaches are not the same (<code>scaling=TRUE</code> does z-score, <code>standardize=TRUE</code> scales to unit central L2 norm), and results will not be identical. Also, using <code>standardize=TRUE</code> scales variables but provides coeffients for the original scale, whereas using scaling=TRUE scales variables in the training set then applies the same scales to the test set.</p>
<div id="summarization-and-plotting" class="section level2">
<h2 class="hasAnchor">
<a href="#summarization-and-plotting" class="anchor"></a>Summarization and plotting</h2>
<p>Cox fit on the continuous risk predictions:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">coxfit.continuous</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/coxph.html">coxph</a></span>(<span class="no">surv.obj</span>~<span class="no">preds</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">coxfit.continuous</span>)</pre></body></html></div>
<pre><code>## Call:
## coxph(formula = surv.obj ~ preds)
## 
##   n= 86, number of events= 24 
## 
##          coef exp(coef) se(coef)     z Pr(&gt;|z|)
## preds 0.01639   1.01653  0.07769 0.211    0.833
## 
##       exp(coef) exp(-coef) lower .95 upper .95
## preds     1.017     0.9837     0.873     1.184
## 
## Concordance= 0.543  (se = 0.057 )
## Likelihood ratio test= 0.04  on 1 df,   p=0.8
## Wald test            = 0.04  on 1 df,   p=0.8
## Score (logrank) test = 0.04  on 1 df,   p=0.8</code></pre>
<p>Dichotomize the cross-validated risk predictions at the median, for visualization:</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">preds.dichot</span> <span class="kw">&lt;-</span> <span class="no">preds</span> <span class="kw">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span>(<span class="no">preds</span>)</pre></body></html></div>
<p>Plot the ROC curve:</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">survivalROC</span>)
<span class="no">nobs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">preds</span>)
<span class="no">cutoff</span> <span class="kw">&lt;-</span> <span class="fl">12</span>
<span class="no">preds.roc</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/pkg/survivalROC/man/survivalROC.html">survivalROC</a></span>(
    <span class="kw">Stime</span> <span class="kw">=</span> <span class="no">beer.survival</span>$<span class="no">os</span>,
    <span class="kw">status</span> <span class="kw">=</span> <span class="no">beer.survival</span>$<span class="no">status</span>,
    <span class="kw">marker</span> <span class="kw">=</span> <span class="no">preds</span>,
    <span class="kw">predict.time</span> <span class="kw">=</span> <span class="no">cutoff</span>,
    <span class="kw">span</span> <span class="kw">=</span> <span class="fl">0.01</span> * <span class="no">nobs</span> ^ (-<span class="fl">0.20</span>)
  )
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(
  <span class="no">preds.roc</span>$<span class="no">FP</span>,
  <span class="no">preds.roc</span>$<span class="no">TP</span>,
  <span class="kw">type</span> <span class="kw">=</span> <span class="st">"l"</span>,
  <span class="kw">xlim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">1</span>),
  <span class="kw">ylim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">1</span>),
  <span class="kw">xlab</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"FP"</span>, <span class="st">"\n"</span>, <span class="st">"AUC = "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(<span class="no">preds.roc</span>$<span class="no">AUC</span>, <span class="fl">3</span>)),
  <span class="kw">lty</span> <span class="kw">=</span> <span class="fl">2</span>,
  <span class="kw">ylab</span> <span class="kw">=</span> <span class="st">"TP"</span>,
  <span class="kw">main</span> <span class="kw">=</span> <span class="st">"LASSO predictions\n ROC curve at 12 months"</span>
)
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span>(<span class="fl">0</span>, <span class="fl">1</span>)</pre></body></html></div>
<div class="figure">
<img src="pensim_files/figure-html/ROCplot-1.png" alt="**Figure 1: ROC plot of cross-validated continuous risk predictions at 12 months.** Note that the predictions are better if you don't randomly select 250 genes to start with!  We only did this to ease the load on the CRAN checking servers." width="700"><p class="caption">
<strong>Figure 1: ROC plot of cross-validated continuous risk predictions at 12 months.</strong> Note that the predictions are better if you don’t randomly select 250 genes to start with! We only did this to ease the load on the CRAN checking servers.
</p>
</div>
</div>
</div>
<div id="getting-coefficients-for-model-fit-on-all-the-data" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-coefficients-for-model-fit-on-all-the-data" class="anchor"></a>Getting coefficients for model fit on all the data</h1>
<p>Finally, we can get coefficients for the model fit on all the data, for future use. Note that nsim should ideally be greater than 1, to train the model using multiple foldings for cross-validation. The output of <code>opt1D</code> or <code>opt2D</code> will be a matrix with one row per simulation. The default behavior in <code><a href="../reference/opt.nested.crossval.html">opt.nested.crossval()</a></code> is to take the simulation with highest cross-validated partial log likelihood (<strong>CVL</strong>), which is the recommended way to select a model from the multiple simulations.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">beer.coefs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/opt1D.html">opt1D</a></span>(
  <span class="kw">setpen</span> <span class="kw">=</span> <span class="st">"L1"</span>,
  <span class="kw">nsim</span> <span class="kw">=</span> <span class="fl">1</span>,
  <span class="kw">response</span> <span class="kw">=</span> <span class="no">surv.obj</span>,
  <span class="kw">penalized</span> <span class="kw">=</span> <span class="no">dat.filt</span>,
  <span class="kw">fold</span> <span class="kw">=</span> <span class="fl">5</span>,
  <span class="kw">maxlambda1</span> <span class="kw">=</span> <span class="fl">5</span>,
  <span class="kw">positive</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
  <span class="kw">standardize</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
  <span class="kw">trace</span> <span class="kw">=</span> <span class="fl">FALSE</span>
)</pre></body></html></div>
<p>We can also include unpenalized covariates, if desired. Note that when keeping only one variable for a penalized or unpenalized covariate, indexing a dataframe like <code>[1]</code> instead of doing <code>[, 1]</code> preserves the variable name. With <code>[, 1]</code> the variable name gets converted to "".</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">beer.coefs.unpen</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="../reference/opt1D.html">opt1D</a></span>(
    <span class="kw">setpen</span> <span class="kw">=</span> <span class="st">"L1"</span>,
    <span class="kw">nsim</span> <span class="kw">=</span> <span class="fl">1</span>,
    <span class="kw">response</span> <span class="kw">=</span> <span class="no">surv.obj</span>,
    <span class="kw">penalized</span> <span class="kw">=</span> <span class="no">dat.filt</span>[-<span class="fl">1</span>],
    <span class="co"># This is equivalent to dat.filt[,-1]</span>
    <span class="kw">unpenalized</span> <span class="kw">=</span> <span class="no">dat.filt</span>[<span class="fl">1</span>],
    <span class="kw">fold</span> <span class="kw">=</span> <span class="fl">5</span>,
    <span class="kw">maxlambda1</span> <span class="kw">=</span> <span class="fl">5</span>,
    <span class="kw">positive</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
    <span class="kw">standardize</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
    <span class="kw">trace</span> <span class="kw">=</span> <span class="fl">FALSE</span>
  )</pre></body></html></div>
<p>Note the non-zero first coefficient this time, due to it being unpenalized:</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">beer.coefs</span>[<span class="fl">1</span>, <span class="fl">1</span>:<span class="fl">5</span>]        <span class="co">#example output with no unpenalized covariates</span></pre></body></html></div>
<pre><code>##             L1            cvl    J04130_s_at U20758_rna1_at    L11672_r_at 
##     4.99896033  -115.19258054     0.00000000     0.09998177    -0.10312924</code></pre>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">beer.coefs.unpen</span>[<span class="fl">1</span>, <span class="fl">1</span>:<span class="fl">5</span>]  <span class="co">#example output with first covariate unpenalized</span></pre></body></html></div>
<pre><code>##             L1            cvl    J04130_s_at U20758_rna1_at    L11672_r_at 
##     4.99906823  -117.76317249    -0.15193598     0.10888431    -0.03394335</code></pre>
</div>
<div id="simulation-of-collinear-high-dimensional-data-with-survival-or-binary-outcome" class="section level1">
<h1 class="hasAnchor">
<a href="#simulation-of-collinear-high-dimensional-data-with-survival-or-binary-outcome" class="anchor"></a>Simulation of collinear high-dimensional data with survival or binary outcome</h1>
<p>The <em>pensim</em> also provides a convenient means to simulation high-dimensional expression data with (potentially censored) survival outcome or binary outcome which is dependent on specified covariates.</p>
<div id="binary-outcome-example" class="section level2">
<h2 class="hasAnchor">
<a href="#binary-outcome-example" class="anchor"></a>Binary outcome example</h2>
<p>First, generate the data. Here we simulate 20 variables. The first 15 (<strong>group “a”</strong>) are uncorrelated, and have no association with outcome. The final five (<strong>group “b”</strong>) have covariance of 0.8 to each other variable in that group. The response variable is associated with the first variable group “b” (<code>firstonly=TRUE</code>) with a coefficient of 2.</p>
<p>Binary outcomes for <span class="math inline">\(n_s= 50\)</span> samples are simulated as a Bernoulli distribution with probability for patient s:</p>
<p><span class="math display">\[\begin{equation}
p_{s} =\frac{1}{1 + exp(-\beta X_{s})}
\end{equation}\]</span></p>
<p>with <span class="math inline">\(\beta_{s,16} = 0.5\)</span> and all other <span class="math inline">\(\beta_{s,i}\)</span> equal to zero.</p>
<p>The code for this simulation is as follows:</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">9</span>)
<span class="no">x</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/create.data.html">create.data</a></span>(
  <span class="kw">nvars</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">15</span>, <span class="fl">5</span>),
  <span class="kw">cors</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">0.8</span>),
  <span class="kw">associations</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">2</span>),
  <span class="kw">firstonly</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">TRUE</span>, <span class="fl">TRUE</span>),
  <span class="kw">nsamples</span> <span class="kw">=</span> <span class="fl">50</span>,
  <span class="kw">response</span> <span class="kw">=</span> <span class="st">"binary"</span>,
  <span class="kw">logisticintercept</span> <span class="kw">=</span> <span class="fl">0.5</span>
)</pre></body></html></div>
<p>Take a look at the simulated data:</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">x</span>)</pre></body></html></div>
<pre><code>##              Length Class      Mode   
## summary        6    data.frame list   
## associations  20    -none-     numeric
## covariance   400    -none-     numeric
## data          21    data.frame list</code></pre>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">x</span>$<span class="no">summary</span></pre></body></html></div>
<pre><code>##   start end cors associations num firstonly
## a     1  15  0.0            0  15      TRUE
## b    16  20  0.8            2   5      TRUE</code></pre>
<p>A simple logistic model fails at variable selection in this case:</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="no">simplemodel</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span>(<span class="no">outcome</span> ~ <span class="no">.</span>, <span class="kw">data</span> <span class="kw">=</span> <span class="no">x</span>$<span class="no">data</span>, <span class="kw">family</span> <span class="kw">=</span> <span class="no">binomial</span>)</pre></body></html></div>
<pre><code>## Warning: glm.fit: algorithm did not converge</code></pre>
<pre><code>## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred</code></pre>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">simplemodel</span>)</pre></body></html></div>
<pre><code>## 
## Call:
## glm(formula = outcome ~ ., family = binomial, data = x$data)
## 
## Deviance Residuals: 
##        Min          1Q      Median          3Q         Max  
## -2.156e-04  -2.100e-08   2.100e-08   4.490e-06   2.110e-04  
## 
## Coefficients:
##               Estimate Std. Error z value Pr(&gt;|z|)
## (Intercept)    394.817  40723.092   0.010    0.992
## a.1             -2.017   8340.192   0.000    1.000
## a.2           1054.116 107650.687   0.010    0.992
## a.3            432.929  41344.208   0.010    0.992
## a.4            187.507  21631.886   0.009    0.993
## a.5           -403.302  39003.798  -0.010    0.992
## a.6           -236.102  20862.924  -0.011    0.991
## a.7           -258.201  23085.549  -0.011    0.991
## a.8              7.364  18988.262   0.000    1.000
## a.9            229.596  21020.703   0.011    0.991
## a.10          -519.764  55465.620  -0.009    0.993
## a.11          -718.759  73461.234  -0.010    0.992
## a.12            94.096  16745.928   0.006    0.996
## a.13          -619.513  65350.574  -0.009    0.992
## a.14           585.887  61785.544   0.009    0.992
## a.15          -271.199  30368.663  -0.009    0.993
## b.1            447.844  49587.944   0.009    0.993
## b.2           -100.723  21001.800  -0.005    0.996
## b.3           -197.254  35209.129  -0.006    0.996
## b.4           1540.346 146783.886   0.010    0.992
## b.5          -1080.818 105901.366  -0.010    0.992
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 6.6406e+01  on 49  degrees of freedom
## Residual deviance: 3.7489e-07  on 29  degrees of freedom
## AIC: 42
## 
## Number of Fisher Scoring iterations: 25</code></pre>
<p>But LASSO does a better job, selecting several of the collinear variables in the “b” group of variables which are associated with outcome:</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="no">lassofit</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="../reference/opt1D.html">opt1D</a></span>(
    <span class="kw">nsim</span> <span class="kw">=</span> <span class="fl">3</span>,
    <span class="kw">nprocessors</span> <span class="kw">=</span> <span class="fl">1</span>,
    <span class="kw">setpen</span> <span class="kw">=</span> <span class="st">"L1"</span>,
    <span class="kw">penalized</span> <span class="kw">=</span> <span class="no">x</span>$<span class="no">data</span>[<span class="fl">1</span>:<span class="fl">20</span>],
    <span class="kw">response</span> <span class="kw">=</span> <span class="no">x</span>$<span class="no">data</span>[, <span class="st">"outcome"</span>],
    <span class="kw">trace</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
    <span class="kw">fold</span> <span class="kw">=</span> <span class="fl">10</span>
  )
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">lassofit</span>)</pre></body></html></div>
<pre><code>##            L1       cvl (Intercept) a.1        a.2 a.3 a.4 a.5         a.6 a.7
## [1,] 4.404100 -32.62007   0.5318400   0 0.04073869   0   0   0 -0.08727491   0
## [2,] 4.992938 -34.51760   0.5204535   0 0.01012467   0   0   0 -0.02152966   0
## [3,] 4.133468 -33.02222   0.5429307   0 0.06228851   0   0   0 -0.11206135   0
##      a.8 a.9        a.10       a.11 a.12 a.13 a.14 a.15        b.1 b.2 b.3
## [1,]   0   0 -0.00379697 -0.3295077    0    0    0    0 0.00000000   0   0
## [2,]   0   0  0.00000000 -0.2793558    0    0    0    0 0.00000000   0   0
## [3,]   0   0 -0.02862559 -0.3483136    0    0    0    0 0.05135596   0   0
##            b.4       b.5
## [1,] 0.2525967 0.1969338
## [2,] 0.2313165 0.1668551
## [3,] 0.2559668 0.1831810</code></pre>
<p>And visualize the data as a heatmap:</p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="no">dat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">x</span>$<span class="no">data</span>[,-<span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="st">"outcome"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">x</span>$<span class="no">data</span>))]))
<span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span>(<span class="no">dat</span>, <span class="kw">ColSideColors</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">x</span>$<span class="no">data</span>$<span class="no">outcome</span> <span class="kw">==</span> <span class="fl">0</span>, <span class="st">"black"</span>, <span class="st">"white"</span>))</pre></body></html></div>
<div class="figure">
<img src="pensim_files/figure-html/heatmap-1.png" alt="**Figure 2: Heatmap of simulated data with binary response.**" width="700"><p class="caption">
<strong>Figure 2: Heatmap of simulated data with binary response.</strong>
</p>
</div>
</div>
<div id="survival-outcome-example" class="section level2">
<h2 class="hasAnchor">
<a href="#survival-outcome-example" class="anchor"></a>Survival outcome example</h2>
<p>We simulate these data in the same way, but with <code>response="timetoevent"</code>. Here censoring is uniform random between times 2 and 10, generating approximately 34% censoring:</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1</span>)
<span class="no">x</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/create.data.html">create.data</a></span>(
  <span class="kw">nvars</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">15</span>, <span class="fl">5</span>),
  <span class="kw">cors</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">0.8</span>),
  <span class="kw">associations</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">0.5</span>),
  <span class="kw">firstonly</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">TRUE</span>, <span class="fl">TRUE</span>),
  <span class="kw">nsamples</span> <span class="kw">=</span> <span class="fl">50</span>,
  <span class="kw">censoring</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>, <span class="fl">10</span>),
  <span class="kw">response</span> <span class="kw">=</span> <span class="st">"timetoevent"</span>
)</pre></body></html></div>
<p>How many events are censored?</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">x</span>$<span class="no">data</span>$<span class="no">cens</span> <span class="kw">==</span> <span class="fl">0</span>) / <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">x</span>$<span class="no">data</span>)</pre></body></html></div>
<pre><code>## [1] 0.42</code></pre>
<p>Kaplan-Meier plot of this simulated cohort:</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">survival</span>)
<span class="no">surv.obj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html">Surv</a></span>(<span class="no">x</span>$<span class="no">data</span>$<span class="no">time</span>, <span class="no">x</span>$<span class="no">data</span>$<span class="no">cens</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/survival/man/survfit.html">survfit</a></span>(<span class="no">surv.obj</span> ~ <span class="fl">1</span>), <span class="kw">ylab</span> <span class="kw">=</span> <span class="st">"Survival probability"</span>, <span class="kw">xlab</span> <span class="kw">=</span> <span class="st">"time"</span>)</pre></body></html></div>
<div class="figure">
<img src="pensim_files/figure-html/simulatedKM-1.png" alt="**Figure 3: Kaplan-Meier plot of survival of simulated cohort.**" width="700"><p class="caption">
<strong>Figure 3: Kaplan-Meier plot of survival of simulated cohort.</strong>
</p>
</div>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></body></html></div>
<pre><code>## R Under development (unstable) (2020-01-28 r77738)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 10 (buster)
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.3.5.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] survivalROC_1.0.3 penalized_0.9-51  survival_3.1-8    pensim_1.2.9     
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.3       knitr_1.28       magrittr_1.5     splines_4.0.0   
##  [5] MASS_7.3-51.5    lattice_0.20-40  R6_2.4.1         rlang_0.4.5     
##  [9] highr_0.8        stringr_1.4.0    tools_4.0.0      grid_4.0.0      
## [13] parallel_4.0.0   xfun_0.12        htmltools_0.4.0  yaml_2.2.1      
## [17] assertthat_0.2.1 digest_0.6.25    rprojroot_1.3-2  pkgdown_1.5.0   
## [21] crayon_1.3.4     Matrix_1.2-18    fs_1.3.2         memoise_1.1.0   
## [25] evaluate_0.14    rmarkdown_2.1    stringi_1.4.6    compiler_4.0.0  
## [29] desc_1.2.0       backports_1.1.5</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-beer_gene-expression_2002">
<p>Beer, David G., Sharon L.R. Kardia, Chiang-Ching Huang, Thomas J. Giordano, Albert M. Levin, David E. Misek, Lin Lin, et al. 2002. “Gene-Expression Profiles Predict Survival of Patients with Lung Adenocarcinoma.” <em>Nat Med</em> 8 (8): 816–24. <a href="https://doi.org/10.1038/nm733">https://doi.org/10.1038/nm733</a>.</p>
</div>
<div id="ref-goeman_l1_2010">
<p>Goeman, Jelle J. 2010. “L1 Penalized Estimation in the Cox Proportional Hazards Model.” <em>Biometrical Journal. Biometrische Zeitschrift</em> 52 (1): 70–84. <a href="https://doi.org/10.1002/bimj.200900028">https://doi.org/10.1002/bimj.200900028</a>.</p>
</div>
<div id="ref-waldron_optimized_2011">
<p>Waldron, Levi, Melania Pintilie, Ming-Sound Tsao, Frances A Shepherd, Curtis Huttenhower, and Igor Jurisica. 2011. “Optimized Application of Penalized Regression Methods to Diverse Genomic Data.” <em>Bioinformatics</em> 27 (24): 3399–3406. <a href="https://doi.org/10.1093/bioinformatics/btr591">https://doi.org/10.1093/bioinformatics/btr591</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Waldron, Levi, Pintilie, Melania, Tsao, {Ming-Sound}, Shepherd, Frances A, Huttenhower, Curtis, Jurisica, Igor.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
