<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>pensim package user guide • pensim</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="pensim package user guide">
<meta property="og:description" content="pensim">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pensim</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/pensim.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/waldronlab/pensim/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>pensim package user guide</h1>
                        <h4 data-toc-skip class="author">Levi
Waldron</h4>
            
            <h4 data-toc-skip class="date">December 06, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/waldronlab/pensim/blob/HEAD/vignettes/pensim.Rmd" class="external-link"><code>vignettes/pensim.Rmd</code></a></small>
      <div class="hidden name"><code>pensim.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This package acts as a wrapper to the <em>penalized</em> R package to
add the following functionality to that package:</p>
<ul>
<li>repeated tuning on different data folds, with parallelization to
take advantage of multiple processors</li>
<li>two-dimensional tuning of the Elastic Net</li>
<li>calculation of cross-validated risk scores through a nested
cross-validation procedure, with parallelization to take advantage of
multiple processors</li>
</ul>
<p>It also provides a function for simulation of collinear
high-dimensional data with survival or binary response.</p>
<p>This package was developed in support of the study by <span class="citation">Waldron et al. (2011)</span>.<br>
This paper contains greater detail on proper application of the methods
provided here. Please cite this paper when using the pensim package in
your research, as well as the penalized package (<span class="citation">Goeman (2010)</span>).</p>
</div>
<div class="section level2">
<h2 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h2>
<p><em>pensim</em> provides example data from a microarray experiment
investigating survival of cancer patients with lung adenocarcinomas
(<span class="citation">Beer et al. (2002)</span>). Load this data and
do an initial pre-filter of genes with low IQR:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://waldronlab.io/pensim/" class="external-link">pensim</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">beer.exprs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">beer.survival</span><span class="op">)</span></span>
<span><span class="co">##select just 100 genes to speed computation, just for the sake of example:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">beer.exprs.sample</span> <span class="op">&lt;-</span> <span class="va">beer.exprs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">beer.exprs</span><span class="op">)</span>, <span class="fl">100</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#</span></span>
<span><span class="va">gene.quant</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">beer.exprs.sample</span>, <span class="fl">1</span>, <span class="va">quantile</span>, probs <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span>
<span><span class="va">dat.filt</span> <span class="op">&lt;-</span> <span class="va">beer.exprs.sample</span><span class="op">[</span><span class="va">gene.quant</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">gene.iqr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">dat.filt</span>, <span class="fl">1</span>, <span class="va">IQR</span><span class="op">)</span></span>
<span><span class="va">dat.filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">dat.filt</span><span class="op">[</span><span class="va">gene.iqr</span> <span class="op">&gt;</span> <span class="fl">0.5</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="va">dat.filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">dat.filt</span><span class="op">)</span></span>
<span><span class="va">dat.filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">dat.filt</span><span class="op">)</span></span>
<span><span class="co">#</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival" class="external-link">survival</a></span><span class="op">)</span></span>
<span><span class="va">surv.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">beer.survival</span><span class="op">$</span><span class="va">os</span>, <span class="va">beer.survival</span><span class="op">$</span><span class="va">status</span><span class="op">)</span></span></code></pre></div>
<p>Note that the expression data are in “wide” format, with one column
per predictor (gene). It is recommended to put covariate data in a
dataframe object, rather than a matrix.</p>
</div>
<div class="section level2">
<h2 id="nested-cross-validation">Nested cross-validation<a class="anchor" aria-label="anchor" href="#nested-cross-validation"></a>
</h2>
<p>Unbiased estimation of prediction accuracy involves two levels of
cross-validation: an outer level for estimating prediction accuracy, and
an inner level for model tuning. This process is simplified by the
opt.nested.crossval function.</p>
<p>It is recommended first to establish the arguments for the penalized
regression by testing on the <em>penalized</em> package functions
<code>optL1</code> (for LASSO), <code>optL2</code> (for Ridge) or
<code>cvl</code> (for Elastic Net). Here we use LASSO. Setting
<code>maxlambda1=5</code> is not a generally recommended procedure, but
is useful in this toy example to avoid converging on the null model.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">penalized</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Welcome to penalized. For extended examples, see vignette("penalized").</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testfit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/penalized/man/cvl.html" class="external-link">optL1</a></span><span class="op">(</span></span>
<span>  response <span class="op">=</span> <span class="va">surv.obj</span>,</span>
<span>  penalized <span class="op">=</span> <span class="va">dat.filt</span>,</span>
<span>  fold <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  maxlambda1 <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  positive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  standardize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  trace <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now pass these arguments to <code><a href="../reference/opt.nested.crossval.html">opt.nested.crossval()</a></code> for
cross-validated calculation and assessment of risk scores, with the
additional arguments:</p>
<ul>
<li>
<code>outerfold</code> and <code>nprocessors</code>: number of folds
for the outer level of cross-validation, and the number of processors to
use for the outer level of cross-validation (see
<code><a href="../reference/opt.nested.crossval.html">?opt.nested.crossval</a></code>)</li>
<li>optFUN and scaling: which optimization function to use
(<code>opt1D</code> for LASSO or Ridge, <code>opt2D</code> for Elastic
Net) - see <code><a href="../reference/opt.splitval.html">?opt.splitval</a></code>.</li>
<li>setpen and nsim: setpen defines whether to do LASSO
(“<strong>L1</strong>”) or Ridge (“<strong>L2</strong>”),
<code>nsim</code> defines the number of times to repeat tuning (see
<code><a href="../reference/opt1D.html">?opt1D</a></code>. <code>opt2D</code> has different required
arguments.)</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">preds</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/opt.nested.crossval.html">opt.nested.crossval</a></span><span class="op">(</span></span>
<span>    outerfold <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    nprocessors <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    <span class="co">#opt.nested.crossval arguments</span></span>
<span>    optFUN <span class="op">=</span> <span class="st">"opt1D"</span>,</span>
<span>    scaling <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    <span class="co">#opt.splitval arguments</span></span>
<span>    setpen <span class="op">=</span> <span class="st">"L1"</span>,</span>
<span>    nsim <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    <span class="co">#opt1D arguments</span></span>
<span>    response <span class="op">=</span> <span class="va">surv.obj</span>,</span>
<span>    <span class="co">#rest are penalized::optl1 arguments</span></span>
<span>    penalized <span class="op">=</span> <span class="va">dat.filt</span>,</span>
<span>    fold <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    positive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    standardize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    trace <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Ideally nsim would be 50, and outerfold and fold would be 10, but the
values below speed computation 200x compared to these recommended
values. Note that here we are using the <code>standardize=TRUE</code>
argument of <code>optL1</code> rather than the <code>scaling=TRUE</code>
argument of opt.splitval. These two approaches to scaling are roughly
equivalent, but the scaling approaches are not the same
(<code>scaling=TRUE</code> does z-score, <code>standardize=TRUE</code>
scales to unit central L2 norm), and results will not be identical.
Also, using <code>standardize=TRUE</code> scales variables but provides
coeffients for the original scale, whereas using scaling=TRUE scales
variables in the training set then applies the same scales to the test
set.</p>
<div class="section level3">
<h3 id="summarization-and-plotting">Summarization and plotting<a class="anchor" aria-label="anchor" href="#summarization-and-plotting"></a>
</h3>
<p>Cox fit on the continuous risk predictions:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coxfit.continuous</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/coxph.html" class="external-link">coxph</a></span><span class="op">(</span><span class="va">surv.obj</span><span class="op">~</span><span class="va">preds</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">coxfit.continuous</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Call:</span></span>
<span><span class="co">## coxph(formula = surv.obj ~ preds)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   n= 86, number of events= 24 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">##          coef exp(coef) se(coef)     z Pr(&gt;|z|)</span></span>
<span><span class="co">## preds 0.01639   1.01653  0.07769 0.211    0.833</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##       exp(coef) exp(-coef) lower .95 upper .95</span></span>
<span><span class="co">## preds     1.017     0.9837     0.873     1.184</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Concordance= 0.543  (se = 0.057 )</span></span>
<span><span class="co">## Likelihood ratio test= 0.04  on 1 df,   p=0.8</span></span>
<span><span class="co">## Wald test            = 0.04  on 1 df,   p=0.8</span></span>
<span><span class="co">## Score (logrank) test = 0.04  on 1 df,   p=0.8</span></span></code></pre>
<p>Dichotomize the cross-validated risk predictions at the median, for
visualization:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">preds.dichot</span> <span class="op">&lt;-</span> <span class="va">preds</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">preds</span><span class="op">)</span></span></code></pre></div>
<p>Plot the ROC curve:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nobs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">preds</span><span class="op">)</span></span>
<span><span class="va">cutoff</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"survivalROC"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span> <span class="va">preds.roc</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">survivalROC</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/survivalROC/man/survivalROC.html" class="external-link">survivalROC</a></span><span class="op">(</span></span>
<span>    Stime <span class="op">=</span> <span class="va">beer.survival</span><span class="op">$</span><span class="va">os</span>,</span>
<span>    status <span class="op">=</span> <span class="va">beer.survival</span><span class="op">$</span><span class="va">status</span>,</span>
<span>    marker <span class="op">=</span> <span class="va">preds</span>,</span>
<span>    predict.time <span class="op">=</span> <span class="va">cutoff</span>,</span>
<span>    span <span class="op">=</span> <span class="fl">0.01</span> <span class="op">*</span> <span class="va">nobs</span> <span class="op">^</span> <span class="op">(</span><span class="op">-</span><span class="fl">0.20</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">preds.roc</span><span class="op">$</span><span class="va">FP</span>,</span>
<span>  <span class="va">preds.roc</span><span class="op">$</span><span class="va">TP</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"l"</span>,</span>
<span>  xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  xlab <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"FP"</span>, <span class="st">"\n"</span>, <span class="st">"AUC = "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">preds.roc</span><span class="op">$</span><span class="va">AUC</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  lty <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"TP"</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"LASSO predictions\n ROC curve at 12 months"</span></span>
<span> <span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="figure">
<img src="pensim_files/figure-html/ROCplot-1.png" alt="**Figure 1: ROC plot of cross-validated continuous risk predictions at 12 months.** Note that the predictions are better if you don't randomly select 250 genes to start with!  We only did this to ease the load on the CRAN checking servers." width="700"><p class="caption">
<strong>Figure 1: ROC plot of cross-validated continuous risk
predictions at 12 months.</strong> Note that the predictions are better
if you don’t randomly select 250 genes to start with! We only did this
to ease the load on the CRAN checking servers.
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="getting-coefficients-for-model-fit-on-all-the-data">Getting coefficients for model fit on all the data<a class="anchor" aria-label="anchor" href="#getting-coefficients-for-model-fit-on-all-the-data"></a>
</h2>
<p>Finally, we can get coefficients for the model fit on all the data,
for future use. Note that nsim should ideally be greater than 1, to
train the model using multiple foldings for cross-validation. The output
of <code>opt1D</code> or <code>opt2D</code> will be a matrix with one
row per simulation. The default behavior in
<code><a href="../reference/opt.nested.crossval.html">opt.nested.crossval()</a></code> is to take the simulation with
highest cross-validated partial log likelihood (<strong>CVL</strong>),
which is the recommended way to select a model from the multiple
simulations.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beer.coefs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/opt1D.html">opt1D</a></span><span class="op">(</span></span>
<span>  setpen <span class="op">=</span> <span class="st">"L1"</span>,</span>
<span>  nsim <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  response <span class="op">=</span> <span class="va">surv.obj</span>,</span>
<span>  penalized <span class="op">=</span> <span class="va">dat.filt</span>,</span>
<span>  fold <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  maxlambda1 <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  positive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  standardize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  trace <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can also include unpenalized covariates, if desired. Note that
when keeping only one variable for a penalized or unpenalized covariate,
indexing a dataframe like <code>[1]</code> instead of doing
<code>[, 1]</code> preserves the variable name. With <code>[, 1]</code>
the variable name gets converted to ““.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beer.coefs.unpen</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/opt1D.html">opt1D</a></span><span class="op">(</span></span>
<span>    setpen <span class="op">=</span> <span class="st">"L1"</span>,</span>
<span>    nsim <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    response <span class="op">=</span> <span class="va">surv.obj</span>,</span>
<span>    penalized <span class="op">=</span> <span class="va">dat.filt</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>    <span class="co"># This is equivalent to dat.filt[,-1]</span></span>
<span>    unpenalized <span class="op">=</span> <span class="va">dat.filt</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>    fold <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    maxlambda1 <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    positive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    standardize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    trace <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Note the non-zero first coefficient this time, due to it being
unpenalized:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beer.coefs</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>        <span class="co">#example output with no unpenalized covariates</span></span></code></pre></div>
<pre><code><span><span class="co">##             L1            cvl    J04130_s_at U20758_rna1_at    L11672_r_at </span></span>
<span><span class="co">##     4.99896033  -115.19258054     0.00000000     0.09998177    -0.10312924</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beer.coefs.unpen</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>  <span class="co">#example output with first covariate unpenalized</span></span></code></pre></div>
<pre><code><span><span class="co">##             L1            cvl    J04130_s_at U20758_rna1_at    L11672_r_at </span></span>
<span><span class="co">##     4.99906823  -117.76317249    -0.15193598     0.10888431    -0.03394335</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="simulation-of-collinear-high-dimensional-data-with-survival-or-binary-outcome">Simulation of collinear high-dimensional data with survival or
binary outcome<a class="anchor" aria-label="anchor" href="#simulation-of-collinear-high-dimensional-data-with-survival-or-binary-outcome"></a>
</h2>
<p>The <em>pensim</em> also provides a convenient means to simulation
high-dimensional expression data with (potentially censored) survival
outcome or binary outcome which is dependent on specified
covariates.</p>
<div class="section level3">
<h3 id="binary-outcome-example">Binary outcome example<a class="anchor" aria-label="anchor" href="#binary-outcome-example"></a>
</h3>
<p>First, generate the data. Here we simulate 20 variables. The first 15
(<strong>group “a”</strong>) are uncorrelated, and have no association
with outcome. The final five (<strong>group “b”</strong>) have
covariance of 0.8 to each other variable in that group. The response
variable is associated with the first variable group “b”
(<code>firstonly=TRUE</code>) with a coefficient of 2.</p>
<p>Binary outcomes for <span class="math inline">\(n_s= 50\)</span>
samples are simulated as a Bernoulli distribution with probability for
patient s:</p>
<p><span class="math display">\[\begin{equation}
p_{s} =\frac{1}{1 + exp(-\beta X_{s})}
\end{equation}\]</span></p>
<p>with <span class="math inline">\(\beta_{s,16} = 0.5\)</span> and all
other <span class="math inline">\(\beta_{s,i}\)</span> equal to
zero.</p>
<p>The code for this simulation is as follows:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">9</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create.data.html">create.data</a></span><span class="op">(</span></span>
<span>  nvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">15</span>, <span class="fl">5</span><span class="op">)</span>,</span>
<span>  cors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.8</span><span class="op">)</span>,</span>
<span>  associations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>  firstonly <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  nsamples <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  response <span class="op">=</span> <span class="st">"binary"</span>,</span>
<span>  logisticintercept <span class="op">=</span> <span class="fl">0.5</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Take a look at the simulated data:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              Length Class      Mode   </span></span>
<span><span class="co">## summary        6    data.frame list   </span></span>
<span><span class="co">## associations  20    -none-     numeric</span></span>
<span><span class="co">## covariance   400    -none-     numeric</span></span>
<span><span class="co">## data          21    data.frame list</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="va">summary</span></span></code></pre></div>
<pre><code><span><span class="co">##   start end cors associations num firstonly</span></span>
<span><span class="co">## a     1  15  0.0            0  15      TRUE</span></span>
<span><span class="co">## b    16  20  0.8            2   5      TRUE</span></span></code></pre>
<p>A simple logistic model fails at variable selection in this case:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simplemodel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="va">outcome</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">x</span><span class="op">$</span><span class="va">data</span>, family <span class="op">=</span> <span class="va">binomial</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: glm.fit: algorithm did not converge</span></span></code></pre>
<pre><code><span><span class="co">## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">simplemodel</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## glm(formula = outcome ~ ., family = binomial, data = x$data)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Deviance Residuals: </span></span>
<span><span class="co">##        Min          1Q      Median          3Q         Max  </span></span>
<span><span class="co">## -8.155e-05  -2.100e-08   2.100e-08   2.100e-08   7.017e-05  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">##               Estimate Std. Error z value Pr(&gt;|z|)</span></span>
<span><span class="co">## (Intercept)    146.045  28566.445   0.005    0.996</span></span>
<span><span class="co">## a.1            205.225  36863.342   0.006    0.996</span></span>
<span><span class="co">## a.2           -350.824  60613.163  -0.006    0.995</span></span>
<span><span class="co">## a.3             26.278   9549.552   0.003    0.998</span></span>
<span><span class="co">## a.4           -221.866  39113.234  -0.006    0.995</span></span>
<span><span class="co">## a.5             95.717  24911.161   0.004    0.997</span></span>
<span><span class="co">## a.6             36.924  12817.964   0.003    0.998</span></span>
<span><span class="co">## a.7           -246.090  44715.891  -0.006    0.996</span></span>
<span><span class="co">## a.8             65.058  13128.095   0.005    0.996</span></span>
<span><span class="co">## a.9           -316.993  58214.882  -0.005    0.996</span></span>
<span><span class="co">## a.10             4.299  20688.797   0.000    1.000</span></span>
<span><span class="co">## a.11          -218.757  39199.270  -0.006    0.996</span></span>
<span><span class="co">## a.12            63.066  20305.078   0.003    0.998</span></span>
<span><span class="co">## a.13            54.937  11733.007   0.005    0.996</span></span>
<span><span class="co">## a.14          -112.590  30198.194  -0.004    0.997</span></span>
<span><span class="co">## a.15           398.687  74940.068   0.005    0.996</span></span>
<span><span class="co">## b.1            498.778  88875.945   0.006    0.996</span></span>
<span><span class="co">## b.2           -297.858  56626.061  -0.005    0.996</span></span>
<span><span class="co">## b.3           -677.344 116466.624  -0.006    0.995</span></span>
<span><span class="co">## b.4            188.575  39169.971   0.005    0.996</span></span>
<span><span class="co">## b.5            699.090 123792.675   0.006    0.995</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## (Dispersion parameter for binomial family taken to be 1)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Null deviance: 6.5342e+01  on 49  degrees of freedom</span></span>
<span><span class="co">## Residual deviance: 6.6338e-08  on 29  degrees of freedom</span></span>
<span><span class="co">## AIC: 42</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Fisher Scoring iterations: 25</span></span></code></pre>
<p>But LASSO does a better job, selecting several of the collinear
variables in the “b” group of variables which are associated with
outcome:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lassofit</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/opt1D.html">opt1D</a></span><span class="op">(</span></span>
<span>    nsim <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    nprocessors <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    setpen <span class="op">=</span> <span class="st">"L1"</span>,</span>
<span>    penalized <span class="op">=</span> <span class="va">x</span><span class="op">$</span><span class="va">data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>,</span>
<span>    response <span class="op">=</span> <span class="va">x</span><span class="op">$</span><span class="va">data</span><span class="op">[</span>, <span class="st">"outcome"</span><span class="op">]</span>,</span>
<span>    trace <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    fold <span class="op">=</span> <span class="fl">10</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">lassofit</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            L1       cvl (Intercept) a.1 a.2 a.3 a.4 a.5 a.6 a.7 a.8 a.9 a.10</span></span>
<span><span class="co">## [1,] 3.436910 -27.56252   0.6389925   0   0   0   0   0   0   0   0   0    0</span></span>
<span><span class="co">## [2,] 4.167477 -27.70431   0.6255205   0   0   0   0   0   0   0   0   0    0</span></span>
<span><span class="co">## [3,] 3.949795 -28.37533   0.6291330   0   0   0   0   0   0   0   0   0    0</span></span>
<span><span class="co">##      a.11 a.12 a.13 a.14      a.15       b.1 b.2 b.3 b.4 b.5</span></span>
<span><span class="co">## [1,]    0    0    0    0 0.0128977 0.9981297   0   0   0   0</span></span>
<span><span class="co">## [2,]    0    0    0    0 0.0000000 0.8915954   0   0   0   0</span></span>
<span><span class="co">## [3,]    0    0    0    0 0.0000000 0.9227009   0   0   0   0</span></span></code></pre>
<p>And visualize the data as a heatmap:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">data</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="st">"outcome"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">data</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html" class="external-link">heatmap</a></span><span class="op">(</span><span class="va">dat</span>, ColSideColors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">outcome</span> <span class="op">==</span> <span class="fl">0</span>, <span class="st">"black"</span>, <span class="st">"white"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="pensim_files/figure-html/heatmap-1.png" alt="**Figure 2: Heatmap of simulated data with binary response.**" width="700"><p class="caption">
<strong>Figure 2: Heatmap of simulated data with binary
response.</strong>
</p>
</div>
</div>
<div class="section level3">
<h3 id="survival-outcome-example">Survival outcome example<a class="anchor" aria-label="anchor" href="#survival-outcome-example"></a>
</h3>
<p>We simulate these data in the same way, but with
<code>response="timetoevent"</code>. Here censoring is uniform random
between times 2 and 10, generating approximately 34% censoring:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create.data.html">create.data</a></span><span class="op">(</span></span>
<span>  nvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">15</span>, <span class="fl">5</span><span class="op">)</span>,</span>
<span>  cors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.8</span><span class="op">)</span>,</span>
<span>  associations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>  firstonly <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  nsamples <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  censoring <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  response <span class="op">=</span> <span class="st">"timetoevent"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>How many events are censored?</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cens</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.42</span></span></code></pre>
<p>Kaplan-Meier plot of this simulated cohort:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival" class="external-link">survival</a></span><span class="op">)</span></span>
<span><span class="va">surv.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">time</span>, <span class="va">x</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cens</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/survfit.html" class="external-link">survfit</a></span><span class="op">(</span><span class="va">surv.obj</span> <span class="op">~</span> <span class="fl">1</span><span class="op">)</span>, ylab <span class="op">=</span> <span class="st">"Survival probability"</span>, xlab <span class="op">=</span> <span class="st">"time"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="pensim_files/figure-html/simulatedKM-1.png" alt="**Figure 3: Kaplan-Meier plot of survival of simulated cohort.**" width="700"><p class="caption">
<strong>Figure 3: Kaplan-Meier plot of survival of simulated
cohort.</strong>
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] penalized_0.9-52 survival_3.4-0   pensim_1.3.6    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] Rcpp_1.0.9        highr_0.9         bslib_0.4.1       compiler_4.2.2   </span></span>
<span><span class="co">##  [5] jquerylib_0.1.4   tools_4.2.2       digest_0.6.30     jsonlite_1.8.3   </span></span>
<span><span class="co">##  [9] evaluate_0.18     memoise_2.0.1     lifecycle_1.0.3   lattice_0.20-45  </span></span>
<span><span class="co">## [13] rlang_1.0.6       Matrix_1.5-1      cli_3.4.1         yaml_2.3.6       </span></span>
<span><span class="co">## [17] parallel_4.2.2    pkgdown_2.0.6     xfun_0.35         fastmap_1.1.0    </span></span>
<span><span class="co">## [21] stringr_1.5.0     knitr_1.41        desc_1.4.2        fs_1.5.2         </span></span>
<span><span class="co">## [25] vctrs_0.5.1       sass_0.4.4        systemfonts_1.0.4 rprojroot_2.0.3  </span></span>
<span><span class="co">## [29] grid_4.2.2        glue_1.6.2        R6_2.5.1          textshaping_0.3.6</span></span>
<span><span class="co">## [33] rmarkdown_2.18    purrr_0.3.5       magrittr_2.0.3    htmltools_0.5.3  </span></span>
<span><span class="co">## [37] survivalROC_1.0.3 splines_4.2.2     MASS_7.3-58.1     ragg_1.2.4       </span></span>
<span><span class="co">## [41] stringi_1.7.8     cachem_1.0.6</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-beer_gene-expression_2002" class="csl-entry">
Beer, David G., Sharon L.R. Kardia, Chiang-Ching Huang, Thomas J.
Giordano, Albert M. Levin, David E. Misek, Lin Lin, et al. 2002.
<span>“Gene-Expression Profiles Predict Survival of Patients with Lung
Adenocarcinoma.”</span> <em>Nat Med</em> 8 (8): 816–24. <a href="https://doi.org/10.1038/nm733" class="external-link">https://doi.org/10.1038/nm733</a>.
</div>
<div id="ref-goeman_l1_2010" class="csl-entry">
Goeman, Jelle J. 2010. <span>“L1 Penalized Estimation in the Cox
Proportional Hazards Model.”</span> <em>Biometrical Journal.
Biometrische Zeitschrift</em> 52 (1): 70–84. <a href="https://doi.org/10.1002/bimj.200900028" class="external-link">https://doi.org/10.1002/bimj.200900028</a>.
</div>
<div id="ref-waldron_optimized_2011" class="csl-entry">
Waldron, Levi, Melania Pintilie, Ming-Sound Tsao, Frances A Shepherd,
Curtis Huttenhower, and Igor Jurisica. 2011. <span>“Optimized
Application of Penalized Regression Methods to Diverse Genomic
Data.”</span> <em>Bioinformatics</em> 27 (24): 3399–3406. <a href="https://doi.org/10.1093/bioinformatics/btr591" class="external-link">https://doi.org/10.1093/bioinformatics/btr591</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Levi Waldron.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
